design/pv-migration.md

```diff
@@ -0,0 +1,86 @@
+
+The problem with the above approach is we can not move the volumes to any existing node as we can set only one label on the node for a given key. So, for any existing node in a cluster, label `openebs.io/nodeid` will already be set so, we can not update it to a new value as the pods already running on it will not able to get scheduled here.
+
+
+## Proposal
+
+### Keys Per ZPOOL
+
+We are proposing to have a key dedicated to ZFS POOL. This key will be used by the ZFS-LocalPV driver to set the label on the nodes where it is present. In this way we can allow the ZFS POOLs to move from any node to any other node as the key is tied to the ZFS POOL as opposed to keeping it per node. We are proposing to have a `openebs.io/<poolname>=true` label on the node where the pool is present. Assuming admins do not have large number of pools on a node, there will be not much label set on a node.
+
+### Migrator
+
+ZFS POOL name should be same across all the nodes for ZFS-LocalPV. So, we have to rename the ZFS POOLs if we are moving it to any existing node. We need a Migrator workflow to update the POOL name in the ZFSVolume object. This will find all the volumes present in a ZFS POOL on that node and updates the ZFSVolume object with the correct PoolName.
+
+**Note:** We can not edit PV volumeAttributes with the new pool name as it is immutable field.
+
+The migrator will look for all the volumes start with the name "pvc-" and will look for corresponding ZFSVolume object and will update it. In case of admin has imported the existing volumes where it is possible that volume name might not start with the name "pvc-", the PoolName needs to be updated manually as the volume was created manually by the admin.
+
+### Workflow
+
+- user will setup all the nodes and setup the ZFS pool on each of those nodes.
+- the ZFS-LocalPV CSI driver will look for all the pools on the node and will set the `openebs.io/<poolname>=true` label for all ZFS POOLs that is present on that node. Let's say node-1 has two pools(say pool1 and pool2) present then the labels will be like this :
+```
+$ kubectl get node pawan-node-1 --show-labels
+NAME           STATUS   ROLES    AGE    VERSION   LABELS
+node-1   Ready    worker   351d   v1.17.4   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=node-1,kubernetes.io/os=linux,node-role.kubernetes.io/worker=true,openebs.io/nodeid=node1,openebs.io/nodename=node-1,openebs.io/pool1=true,openebs.io/pool2=true
+```
+- If we are moving the pool1 from node1 to node2, then there are two cases here :-
+
+#### 1. if node2 is a fresh node
+
+- we can simply import the pool and restart the ZFS-LocalPV driver to make it aware of that pool to set the corresponding node topology
+- the ZFS-LocalPV driver will look for `openebs.io/pool1=true` and removes the label from the nodes where pool is not present 
```

Comment:
How can we ensure it will unlabel the correct node in case multiple nodes have the same pool name?

Reply:
It is openebs.io/pool-guid. So node agent knows which pools it own. And can take the decision.